### PR TITLE
convert to typescript, add noun->jso helpers

### PR DESCRIPTION
**Pull request by polwex**
As for UF grant https://urbit.org/grants/eyre-noun-channels

Converted whole library to TypeScript, except some small parts of the nock compiler (which could be doable but I would need to talk to the original author to see).

Also removed `jsbn` dependency and replaced with JavaScript's now built-in `bigint`. 

Tested well in my handwritten tests inspired by the old tests but proper, modern, thorough testing setup should be done. My nock-foo isn't good enough so I would appreciate suggestions on what's good data to test with.

### CODE REVIEW

File: src/bigint.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  const paddedHexString = hexString.length % 2 === 0 ? hexString : '0' + hexString;
+  const arrayLength = paddedHexString.length / 2;
+  const int8Array = new Int8Array(arrayLength);
+  // const int8Array = new Int8Array(arrayLength + 1);
+  // int8Array[0] = bigInt >= 0n ? 0 : -1;
+  for (let i = 0; i < paddedHexString.length; i += 2) {
+    const hexSubstring = paddedHexString.slice(i, i + 2);
+    const signedInt = (parseInt(hexSubstring, 16) << 24) >> 24;
+    // int8Array[(i / 2)+ 1] = signedInt;
+    int8Array[(i / 2)] = signedInt;
```
Comment (Fang-) [inline src/bigint.ts]: Are the commented out lines incorrect, off-by-one versions of the implementation? Best to clean these up.

File: src/noun-helpers.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+function dwim(a: Atomizable, b: Atomizable): Cell<Atom, Atom>;
+function dwim(
+  a: Atomizable,
+  b: Atomizable,
+  c: Atomizable
+): Cell<Atom, Cell<Atom, Atom>>;
+function dwim(a: Atomizable, ...b: any[]): Cell<Atom, Cell<Noun, Noun>>;
+function dwim(...a: any[]): Cell<Noun, Noun>;
+// implementation
+function dwim(...args: any[]): Noun {
```
Comment (Fang-) [inline src/noun-helpers.ts]: I guess the only real "argument type -> result type" inference we're gonna get in typescript is the one we manually specify, hence all the different cases here? Means arbitrary more-complex cases won't have nice type inference, but I wonder how often that comes up in practice. Type-checking should probably happen in the js layers of code, and not at or after the conversion of that js object into a noun...

All that to say: if this is the best we can get, then this is fine.

File: jest.config.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export default {
+  maxWorkers: 1,
+  testPathIgnorePatterns: ["/node_modules/", "\\.ts$", "\\.tsx$"],
```
Comment (Fang-) [inline jest.config.js]: This seems to be excluding the `*.test.ts` files also, which may not be what's intended? But even with this filter removed, trying to actually run the `index.test.ts` file doesn't succeed, "unexpected token" shenanigans that aren't trivially resolvable.

**Review by Fang- — APPROVED** [review]
This is great! Some small comments, mostly just minor cleanup and non-actionable thoughts. Also I can't seem to run the tests. I do believe they pass, iirc you mentioned running them before, but do want to see so on my own machine as well.

I'm gonna want @arthyn or someone else to weigh in on what the modern-day ts config & testing practices are so we can get those consistent with the other libraries/projects, but I don't expect anything major to fall out of that. As it stands right now, this is already a good improvement, and the addition of the `noun-helpers.ts` makes this very useful for a nouns-over-http world.

File: jest.config.js (hunk 3 above)
Comment (polwex) [reply jest.config.js]: Yes, I had more success building the app (just run `tsc`) and then run `npm run test` to run the compiled test file `dist/index.test.js`.

If @arthyn or someone else has any advice on how to run the ts file itself I'm happy to fix it.

File: src/noun-helpers.ts (hunk 2 above)
Comment (polwex) [reply src/noun-helpers.ts]: I added the function overloading to `dwim()` mostly to properly type the functions in `serial.ts`.

Of all the library by the far the most time and effort I put in trying to have proper recursive typing of Cell nouns for this part but this is the best I could do. Any other approach just hit a wall in the inference not going beyond `Cell<Noun, Noun>`.

**Comment by Fang-** [comment]
Initial recommendation for the "common practices" bit is to look at what [urbit/aura-js](https://github.com/urbit/aura-js) is doing. Looks like it also uses jest for testing, but alongside dts for managing/compiling(?) the typescript. Not too far off.

**Comment by polwex** [comment]
> Initial recommendation for the "common practices" bit is to look at what [urbit/aura-js](https://github.com/urbit/aura-js) is doing. Looks like it also uses jest for testing, but alongside dts for managing/compiling(?) the typescript. Not too far off.

Thanks for the recommendation, I added `dts` and it's running nicely. You can run the tests now without building the module.

**Review by arthyn — COMMENTED** [review]
My only comment here is that urbit/aura-js uses big-integer and not the native version unsure if we should change that or use that here tbh, but it would probably be best if they matched.

**Comment by Fang-** [comment]
From the big-integer npm page:

> Update (December 2, 2018): [BigInt is being added as a native feature of JavaScript](https://tc39.github.io/proposal-bigint/). This library now works as a polyfill: if the environment supports the native BigInt, this library acts as a thin wrapper over the native implementation.

So I think that means we can do a drop-in replacement either direction if we want. I think going full-native is better in the long run.

**Comment by Fang-** [comment]
Long overdue. Let's get this merged. We should deploy a new version soon.

**Pull request merged**